streams/src/test/java/org/apache/kafka/streams/integration/KTableSourceTopicRestartIntegrationTest.java

```diff
@@ -0,0 +1,250 @@
+            // the state restore listener will append two records to the log
+            streamsOne.setGlobalStateRestoreListener(new UpdatingSourceTopicOnRestoreStartStateRestoreListener());
+            streamsOne.start();
+
+            produceKeyValues("f", "g", "h");
+
+            assertNumberValuesRead(readKeyValues, 8, "Table did not get all values after restart");
+
+        } finally {
+            streamsOne.close(5, TimeUnit.SECONDS);
+        }
+    }
+
+    @Test
+    public void shouldRestoreAndProgressWhenTopicNotWrittenToDuringRestoration() throws Exception {
+
+        try {
+            streamsOne = new KafkaStreams(streamsBuilder.build(), STREAMS_CONFIG);
+            streamsOne.start();
+
+            produceKeyValues("a", "b", "c");
+
+            assertNumberValuesRead(readKeyValues, 3, "Table did not read all values");
+
+            streamsOne.close();
+            streamsOne = new KafkaStreams(streamsBuilder.build(), STREAMS_CONFIG);
+            streamsOne.start();
+
+            produceKeyValues("f", "g", "h");
+
+            assertNumberValuesRead(readKeyValues, 6, "Table did not get all values after restart");
+
+        } finally {
+            streamsOne.close(5, TimeUnit.SECONDS);
+        }
+    }
+
+    private void assertNumberValuesRead(final Map<String, String> valueMap,
+                                        final int expectedCount,
+                                        final String errorMessage) throws InterruptedException {
+
+        TestUtils.waitForCondition(new TestCondition() {
+            @Override
+            public boolean conditionMet() {
+                return valueMap.size() == expectedCount;
+            }
+        }, errorMessage);
+
+    }
+
+    private void produceKeyValues(String... keys) throws ExecutionException, InterruptedException {
+        final List<KeyValue<String, String>> keyValueList = new ArrayList<>();
+
+        for (final String key : keys) {
+            keyValueList.add(new KeyValue<>(key, key + "1"));
+        }
+
+        IntegrationTestUtils.produceKeyValuesSynchronously(SOURCE_TOPIC,
+                                                           keyValueList,
+                                                           PRODUCER_CONFIG,
+                                                           time);
+    }
+
+    private class UpdatingSourceTopicOnRestoreStartStateRestoreListener implements StateRestoreListener {
+
+        @Override
+        public void onRestoreStart(final TopicPartition topicPartition,
+                                   final String storeName,
+                                   final long startingOffset,
+                                   final long endingOffset) {
+            try {
+                // need two values appended to log to trigger issue
+                produceKeyValues("d", "e");
+            } catch (ExecutionException | InterruptedException e) {
+                e.printStackTrace();
```

Comment:
Should we set a boolean flag that we can check in the main test method? Otherwise this might be unnoticed.

Reply:
I'd rather throw a `RuntimeException` as an exception here means the test will fail.

Reply:
Not sure. We would catch the RuntimeException and rethrow as StreamsException and thus the StreamThread finally dies. This might make the test fail/timeout ? Or would we need to register an uncaught exception handler? This this case, we don't get any advantage as we need to set a flag there and check. The "issue" is, that throwing here will not bubble out into the test method itself.